### PR TITLE
[Pubgrub] Further improvements

### PR DESCRIPTION
**Pull request by kiliankoe**
Includes further fixes for deterministic tests, non-existent version handling in decision making and checking for satisfaction on more complex cases.

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Comment by shahmishal** [comment]
CI is currently offline

**Comment by shahmishal** [comment]
@swift-ci smoke test

**Comment by shahmishal** [comment]
@swift-ci smoke test macOS

File: Sources/PackageGraph/Pubgrub.swift [hunk 1]
```diff
@@ -1103,14 +1123,15 @@ public final class PubgrubDependencyResolver<
...
                         return Incompatibility(terms, root: root!, cause: .dependency(package: candidate.package))
-                }
+                    }
                 depIncompatibilities.forEach { add($0, location: .decisionMaking) }
 
+                let tmp = PartialSolution(assignments: solution.assignments)
+                tmp.decide(candidate.package, atExactVersion: version)
+                // Check if this decision would result in a conflict when added.
+                // If so, we try the next earlier version instead.
+                #warning("Why is this depIncompatibilities and not incompatibilities[candidate.package]?")
```
Comment (aciidb0mb3r) [inline Sources/PackageGraph/Pubgrub.swift]: @kiliankoe can you remove this warning and make a new PR?

Comment (aciidb0mb3r) [reply Sources/PackageGraph/Pubgrub.swift]: nvm: https://github.com/apple/swift-package-manager/pull/1978

Comment (kiliankoe) [reply Sources/PackageGraph/Pubgrub.swift]: Ah sry, this shouldn't have been staged 🤦‍♂️

**Pull request merged**